src/apply.ts

```diff
@@ -26,154 +24,45 @@ interface ReadOptions {
 	application: number;
 	authToken: string;
 	stream: stream.Readable;
-	force?: boolean;
-}
-
-function generateUniqueKey() {
-	return uuidv4().replace(/-/g, '');
-}
-
-async function backfillLowerRevisions(
-	sdk: SDK.BalenaSDK,
-	application: number,
-	currentDateIso: string,
-	manifest: ReleaseManifest,
-): Promise<void> {
-	const [highestSameSemverLocalReleases] = await sdk.pine.get({
-		resource: 'release',
-		options: {
-			$top: 1,
-			$select: ['id', 'revision'],
-			$filter: {
-				belongs_to__application: application,
-				semver_major: manifest.semver_major,
-				semver_minor: manifest.semver_minor,
-				semver_patch: manifest.semver_patch,
-				revision: { $ne: null },
-			},
-			$orderby: 'revision desc',
-		},
-	});
-
-	const version = `${manifest.semver_major}.${manifest.semver_minor}.${manifest.semver_patch}`;
-	for (
-		//
-		let revision =
-			highestSameSemverLocalReleases == null
-				? 0
-				: // we have already $filtered for this so ! is fine
-					highestSameSemverLocalReleases.revision! + 1;
-		revision < (manifest.revision ? manifest.revision : 0);
-		revision++
-	) {
-		await sdk.pine.post({
-			resource: 'release',
-			body: {
-				belongs_to__application: application,
-				status: 'failed',
-				start_timestamp: currentDateIso,
-				end_timestamp: currentDateIso,
-				update_timestamp: currentDateIso,
-				source: manifest.source,
-				composition: manifest.composition,
-				commit: generateUniqueKey(),
-				contract: manifest.contract,
-				semver: revision === 0 ? version : `${version}+rev${revision}`,
-			},
-		});
-		await new Promise((resolve) => setTimeout(resolve, 1000));
-	}
 }
 
 export async function apply(options: ReadOptions): Promise<number> {
 	// FIXME: clone release timestamps when the API already allows it
 	const currentDateIso = new Date(Date.now()).toISOString();
 
-	const bundle = await resourceBundle.read<ReleaseManifest>(
+	const bundle = await resourceBundle.read<SDK.Release>(
 		options.stream,
 		'io.balena.release',
 	);
 
+	// TODO: validate if manifest version is compatible with SDK
+
+	if (!bundle.manifest) {
+		throw new Error('Release bundle should have a manifest.');
+	}
+
 	const sdk = SDK.getSdk({
 		apiUrl: options.apiUrl,
 		dataDirectory: false,
 	});
 	await sdk.auth.loginWithToken(options.authToken);
 
-	// TODO: validate manifest
-	// Check if a release with the same version and revision exists
-	const [existingRelease] = await sdk.pine.get({
+	const existingReleases = await sdk.pine.get({
 		resource: 'release',
 		options: {
-			$select: [
```

Comment:
What if there are thousands of releases? This query would fetch them all and this is extremely wasteful. Using “$select” with raw Pine queries is good practice and in this case where you don’t care about the actual data, you can just “$select: id”. Also, given you only care about whether at least one release exists, you can limit your query to fetch only one row with “$top: 1”.